.travis.yml

```
@@ -1,21 +1,69 @@
 language: d
-sudo: false
+
+# seems the default 'xenial' isn't working well with mariadb (at some point we need to update to bionic)
+dist: precise
 
 addons:
   mariadb: '10.1'
   apt:
+    update: true
     packages: [ libevent-dev ]
+  homebrew:
+    brewfile: true
 
 install: $DMD -ofci_setup ci_setup.d && ./ci_setup
 script: ./run_tests --mode=combined
 
-matrix:
+jobs:
   include:
+    - stage: Test with all dependencies updated (dub.selections.json is deliberately kept old)
+      d: dmd
```

I should document this, but the omission of "latest version" tags like "dmd", "dmd-beta", "ldc" is intentional because it destroys reproducability of builds since these labels deliberately change meaning over time. In general I regard them as poor, hacky substitutes for what we REALLY need: a bot that submits PRs for each new compiler version (this would NOT break the concept of reproducible builds).

However, I *am* in favor of adding jobs for "dmd", "dmd-beta", "ldc", etc as long as they're marked "allowed-failure" (which makes them purely informative). That would definitely be a worthwhile enhancement.

---

I don't think latest stable DMD and LDC should allow failures. Dmd-beta is fair enough. The fact that I made this the first stage was so that the build can fail fast if not working with current DMD. That should be the priority

---

I think Nick is right, we should have a "failure allowed" version. There is no reason to prevent all development on this project while bugs are fixed in the compiler/phobos. It then becomes a "drop everything" situation. When new versions are released, if the job just works, the new compiler version can be added to the list (if needed).

But I do think we should have something that runs the latest and allows failure, because it would be good to have that warning that something isn't working.

---

that's the purpose of having dmd-beta and ldc-beta though. That's where you should allow failures and create a bug ticket if compiler changes are going to break things. When the compiler is released is too late. If dmd and ldc stable are allowed to fail then what would be the point of building with dmd-beta at all?

Also, unless the CI is maintained it will again fall in to the situation as your master branch in which the latest compiler being tested is dmd-2.080.0, that's a year old! The latest stable is 2.088.0 so in the meantime your codebase could've been broken on any comiler release in that time.

---

The issue is simply that you are then allowing an external project to break development on this project. Why should any breaking changes (or bugs in the release that didn't get caught during beta) make it so PRs here cannot pass? In addition, even if you DO find a bug in beta, there is no guarantee it gets fixed before released to stable.

---

If the codebase stops working with the latest stable D compiler it's broken anyway and it should fail CI. Allowing the build to continue passing won't resolve anything, it just makes the situation worse. Besides we're not talking an external dependency on some random github code, it's the language itself. 

Since this pull request was opened there's already been another dmd release. If I hard code dmd 2.088.0 to be the most recent in the travis file your codebase could be completely broken by language changes in the future and there would be no failing build to flag the problem. Sure you'd have green builds but so what? Better to have a master branch that compiles with the current version of the language.

---

If a D release breaks mysql-native's master branch because of legitimate changes, we need a new release to fix that, and that work can happen in _parallel_ with existing work that's being done. The issue is creating "drop everything" problem that is fully triggered by outside projects. There have been several instances where D incorrectly broke major projects, and Phobos/dmd was changed _back_ to fix the problem, because the release wasn't vetted properly. During that time where the projects are waiting for a point release to fix it, there's no reason to halt all development in the D world.